File: 19_remove_nth_node_from_end_of_list/go/main.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	front, back := head, head
	for i := 0; i < n; i++ {
		front = front.Next
	}
	for front != nil && front.Next != nil {
		front, back = front.Next, back.Next
	}
	if back == head {
		if front != nil {
			back.Next = back.Next.Next
		} else {
			head = head.Next
		}
	} else {
		if back.Next != nil {
			back.Next = back.Next.Next
		}
	}
	
	return head
}

func printLink(head *ListNode) {
	p := head
	for p != nil {
		fmt.Printf("%d ", p.Val)
		p = p.Next
	}
	fmt.Println()
}

func main() {
	node1 := &ListNode{Val: 1, Next: nil}
	node2 := &ListNode{Val: 2, Next: nil}
	node1.Next = node2
	/*node3 := &ListNode{Val: 3, Next: nil}
	node2.Next = node3
	node4 := &ListNode{Val: 4, Next: nil}
	node3.Next = node4
	node5 := &ListNode{Val: 5, Next: nil}
	node4.Next = node5*/

	printLink(node1)
	node1 = removeNthFromEnd(node1, 2)
	printLink(node1)
}

File: 19_remove_nth_node_from_end_of_list/main.py
# -*- coding: utf-8 -*-


class ListNode(object):

    def __init__(self, val):
        self.val = val
        self.next = None


class Solution(object):

    def remove_nth_from_end(self, head, n):
        front = back = head
        for i in range(n):
            front = front.next
        while front is not None and front.next is not None:
            front, back = front.next, back.next
        if back == head:
            if front is not None:
                back.next = back.next.next
            else:
                head = head.next
        else:
            back.next = back.next.next if back.next is not None else None

        return head


def print_link(head):
    p = head
    while p is not None:
        print(p.val),
        p = p.next
    print('')
        
    
if __name__ == '__main__':
    node1 = ListNode(1)
    node2 = ListNode(2)
    node1.next = node2
    # node3 = ListNode(3)
    # node2.next = node3
    # node4 = ListNode(4)
    # node3.next = node4
    # node5 = ListNode(5)
    # node4.next = node5
    # print_link(node1)

    s = Solution()
    node1 = s.remove_nth_from_end(node1, 1)
    print_link(node1)
